File: backend/src/wechat_client.py
from __future__ import annotations

import json
from urllib.parse import urlencode
from urllib.request import urlopen


def exchange_code(app_id: str, app_secret: str, code: str) -> dict:
    query = urlencode(
        {
            "appid": app_id,
            "secret": app_secret,
            "js_code": code,
            "grant_type": "authorization_code",
        }
    )
    with urlopen(f"https://api.weixin.qq.com/sns/jscode2session?{query}", timeout=10) as response:
        payload = json.loads(response.read().decode("utf-8"))

    if payload.get("errcode"):
        raise RuntimeError(payload.get("errmsg") or f"WeChat auth failed: {payload['errcode']}")

    return payload

File: backend/src/wechatClient.js
'use strict'

const https = require('node:https')

function exchangeCode({ appId, appSecret, code }) {
  const url = new URL('https://api.weixin.qq.com/sns/jscode2session')
  url.searchParams.set('appid', appId)
  url.searchParams.set('secret', appSecret)
  url.searchParams.set('js_code', code)
  url.searchParams.set('grant_type', 'authorization_code')

  return new Promise((resolve, reject) => {
    https.get(url, response => {
      let body = ''
      response.setEncoding('utf8')
      response.on('data', chunk => {
        body += chunk
      })
      response.on('end', () => {
        try {
          const parsed = JSON.parse(body)
          if (parsed.errcode) {
            reject(new Error(parsed.errmsg || `WeChat auth failed: ${parsed.errcode}`))
            return
          }
          resolve(parsed)
        } catch (error) {
          reject(error)
        }
      })
    }).on('error', reject)
  })
}

module.exports = {
  exchangeCode
}

